### ISSUE TITLE
Magento discards most New Relic cron transactions

### ISSUE THREAD

**Issue by convenient** [magento/magento2#34108]
# Summary

Magento is discarding all new relic transaction logging/runtime information/errors for every cron job after the first one executed per process.

For example if you had three pending jobs `job1`, `job2`, and `job3` you would only have a transaction for `job1` and no info at all in new relic for `job2` and `job3`.

This is a critical issue IMO as it reduces the effectiveness of our ability to support the Magento instance. I assumed I was getting all relevant data in new relic but I am actually missing performance analytics as well as possible errors that are not being tracked in new relic. This affected me debugging downtime issues affecting a website.

# EDIT 2025-09-03

This has been fixed in 2.4.8 by https://github.com/magento/magento2/commit/76f8b020b4cb5303957f2e9a43ad9cf88baf299f

# Cause

This issue was introduced in 2.4.0 by 5819ded0ed228e0de105000d92da26531c07785a

In that commit there were attempts to 
1. stop logging the `cron:run` command as a command (the `isCommandSkipped` logic) 
2. set the transaction name for each pending job processed in the queue.

The current implementation works a bit like so, starting at `Magento\Cron\Observer\ProcessCronQueueObserver::execute` 

https://github.com/magento/magento2/blob/4c36116dcf878e127059d9be9566a119783583f2/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L273

calls

https://github.com/magento/magento2/blob/4c36116dcf878e127059d9be9566a119783583f2/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L825

calls

https://github.com/magento/magento2/blob/4c36116dcf878e127059d9be9566a119783583f2/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L858

This is the part of the process that actually executes the job, it does so like this

https://github.com/magento/magento2/blob/4c36116dcf878e127059d9be9566a119783583f2/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L350-L377

You can see that each execution of a job is wrapped with `startProfiling` and `stopProfiling`. 

These functions have plugins in them to set `newrelic_name_transaction` and `newrelic_end_transaction`

https://github.com/magento/magento2/blob/4c36116dcf878e127059d9be9566a119783583f2/app/code/Magento/NewRelicReporting/Plugin/StatPlugin.php#L62-L73

https://github.com/magento/magento2/blob/4c36116dcf878e127059d9be9566a119783583f2/app/code/Magento/NewRelicReporting/Plugin/StatPlugin.php#L84-L93

The issue is that we call `newrelic_end_transaction` but nowhere in the codebase is there another call to `newrelic_start_transaction`, this is required to track subsequent jobs.

> For task queue managers, you can use newrelic_end_transaction() and newrelic_start_transaction() together to manually separate transactions
- https://docs.newrelic.com/docs/agents/php-agent/php-agent-api/newrelic_start_transaction/#stop-and-start

This is the major bug, without this call to restart the transaction all subsequent data is not tracked and reported to New Relic.

There is a second quality of life issue. Due to the location where we set the initial transaction name we end up with two traces per cron execution, one of them is flagged as `unknown` which encompasses everything up until we call `newrelic_name_transaction` for the first time, so basically when you see `unknown` on the CLI New Relic data you can think of it as "the time it takes to boot up `cron:run` and get to the point where you run the first actual job".

In effect what we have is the following 
```php
<?php
sleep(10); // cron:run loader
foreach ($i=1; $i<=3; $i++) {
    newrelic_name_transaction("job$i");
    sleep(10);
    newrelic_end_transaction();
}
```

Results
- `unknown` transaction containing the booting of `cron:run`
- `job1` transaction only

When we need something like 

```php
<?php
newrelic_name_transaction("cron:run loader");
sleep(10); // cron:run loader
newrelic_end_transaction();
newrelic_start_transaction(ini_get("newrelic.appname"));
foreach ($i=1; $i<=3; $i++) {
    newrelic_name_transaction("job$i");
    sleep(10);
    newrelic_end_transaction();
    newrelic_start_transaction(ini_get("newrelic.appname"));
}
```

Results
- `cron:run loader` transaction containing the booting of `cron:run`
- `job1` transaction
- `job2` transaction
- `job3` transaction

# Hotfix

For anyone looking for an urgent patch to get their crons all properly logged in New Relic the following seems to be working for me. This ensures we always have a transaction running until this process terminates.

```diff
--- Model/NewRelicWrapper.php	2021-01-07 21:56:58.000000000 +0000
+++ Model/NewRelicWrapper.php	2021-09-17 09:11:31.000000000 +0100
@@ -79,6 +79,7 @@
     {
         if ($this->isExtensionInstalled()) {
             newrelic_end_transaction($ignore);
+            newrelic_start_transaction(ini_get("newrelic.appname"));
         }
     }
```

# Proper solution

I have some rough ideas but I'll leave this for someone in Magento to pick up. I cant see any trivial way of doing what I propose as we'd need to couple the Cron module to New Relic reporting module a bit tighter.

Firstly we can give the `cron:run` its old transaction name back by removing this.
https://github.com/magento/magento2/blob/5819ded0ed228e0de105000d92da26531c07785a/app/code/Magento/NewRelicReporting/etc/di.xml#L49-L55

Then before we call `startProfiling` in `ProcessCronQueueObserver.php` we need to end the current new relic transaction, then start a new one.  

It would also feel weird for `NewRelicWrapper::endTransaction` to have the responsibility of `newrelic_start_transaction`, but I wonder whether adding a new `NewRelicWrapper::startTransaction` would break backwards compatibility if its a new public method 🤔 But the breakage here is so severe that it makes the cron data for new relic practically without use so something needs to be addressed here.

---

Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-Hotel** [comment]
Hello @convenient,

Are you still facing the issue?

Can you please tell us some more information on this issue like:
- Information on your environment
- Steps to reproduce
- Expected and actual results

Thanks

**Comment by convenient** [comment]
@engcom-Hotel this is not environment specific, any Magento installation with New Relic should do it 

The steps to reproduce and the expected and actual are already in the description.

If you are not familiar with New Relic this may need escalated internally your side.

**Comment by convenient** [comment]
@engcom-Hotel 

**System**
- Magento 2.4 (any)
- New Relic configured
- Crons running

**Expected result**
All cron jobs that run have a corresponding new relic transaction
>Results
> 
> - cron:run loader transaction containing the booting of cron:run
> - job1 transaction
> - job2 transaction
> - job3 transaction

**Actual**
Only the first cron run per process has a new relic transaction
> Results
> 
> - unknown transaction containing the booting of cron:run
> - job1 transaction only

**Additional**

I cant tell you exactly what cron transactions you wont be seeing, but it will be most of them. it depends on the roll of the dice in your cron scheduler which I can't really control or predict.

You don't really have to "do" anything to reproduce this issue other than have your system set up with crons running and new relic.

If you get this set up and running and you will see transactions in new relic for your crons, you should be able to apply the hotfix patch I mentioned in my description above and you will then see an increase in the new relic throughput (as more transactions are logged) as well as the full list of all cron transactions. That should help you see what kind of data is missing.

**Comment by convenient** [comment]
@engcom-Hotel can I please also have the appropriate labels applied to this issue?

It's not flagged that it's a https://github.com/magento/magento2/labels/Partner%3A%20Ampersand issue

**Comment by convenient** [comment]
@engcom-Hotel can you please respond on the above?

**Comment by engcom-Hotel** [comment]
Hello @convenient,

Thanks for the report and collaboration!

We have tried to reproduce the issue in Magento 2.4-develop with and without applying the patches. But it seems that the issue is not reproducible for us. Please find the below screenshot for reference:

### New Relic Configuration
![image](https://user-images.githubusercontent.com/51681618/167116451-04294db7-e218-4bc9-a52c-e5c73ddcbc30.png)

### Cron installed
![image](https://user-images.githubusercontent.com/51681618/167116503-82d5290c-a796-452c-9f2b-ae0d6aebd20b.png)

### Before applying the patch
![image](https://user-images.githubusercontent.com/51681618/167116571-f2304838-c85f-4368-80fe-e1889c793041.png)

### After applying the patch
![image](https://user-images.githubusercontent.com/51681618/167116635-4a8e876e-1c9e-458e-80aa-b5683a89279d.png)

In both conditions, we are not able to see the cron job transaction. Let us know in case we have missed anything.

Thanks

**Comment by engcom-Hotel** [comment]
Hello @convenient,

Thanks for the reply!

We have checked, cron are working for us. But still, we are unable to see much transaction in NewRelic. But we are still working on it and get back to you soon.

Thanks

**Comment by engcom-Hotel** [comment]
Hello @convenient,

We have tried to reproduce the issue in Magento 2.4-develop branch with the mentioned steps and the issue is reproducible for us. The NewRelic is showing Cron jobs as `unknown` transactions instead of proper names. Please find below the screenshot for reference:

![image](https://user-images.githubusercontent.com/51681618/168281902-edc4e739-90d4-4464-b800-b55ef87bac41.png)

Hence confirming the issue.

Thanks

**Comment by jerrylopez** [comment]
Any updates on this issue?

**Comment by convenient** [comment]
@jerrylopez FWIW I've had my hotfix composer patched into our instances for over a year now with no ill effects.

**Comment by imswapnil** [comment]
@engcom-Hotel Was there any progress on the issue.
@convenient for some reason the hotfix change in NewRelicWrapper.php you provided did not work for me on 2.4.5-P1.

**Comment by hostep** [comment]
It seems like a fix was pushed to `2.4-develop` branch today: https://github.com/magento/magento2/commit/76f8b020b4cb5303957f2e9a43ad9cf88baf299f

**Comment by engcom-Hotel** [comment]
Thanks @convenient, for the feedback. We are closing this issue as per the above comment.
